### PR TITLE
feat: add precise type to `predicate` in `invalidate`

### PR DESCRIPTION
**Pull request by SrBrahma**
Re https://github.com/trpc/trpc/issues/3496#issuecomment-1369238236
Closes https://github.com/trpc/trpc/issues/3545

## 🎯 Changes

Adds the corresponding input type to `predicate` in `invalidate` on `trpc.useContext().router.procedure`.

## Before
image

## After
image

Notice that the `queryKey` has the right type in [0] and [1]. `type` can be `query` or `infinite`, depending on the procedure.

* For reference, this is how the procedure is defined:

image

## ✅ Checklist

- [X] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/react-query/src/internals/getArrayQueryKey.ts [hunk 1]
```diff
@@ -5,6 +5,11 @@ export type QueryKey = [
   { input?: unknown; type?: Exclude<QueryType, 'any'> }?,
 ];
 
+export type QueryKeyKnown<TInput, TType extends Exclude<QueryType, 'any'>> = [
+  string[],
+  { input: TInput; type: TType },
+];
```
Comment (SrBrahma) [inline packages/react-query/src/internals/getArrayQueryKey.ts:11]: I had to add this as `QueryKey` is not generic and I wasn't able to make it generic as it considers the second array item to be optional. Changes on it are welcome!

**Comment by SrBrahma** [comment]
It seems that the queryKey[1] can be undefined if the query has `enabled=false`. Is this true? If so, I need to add `| undefined` to queryKey[1].

**Comment by nicolassanmar** [comment]
As the creator of https://github.com/trpc/trpc/issues/3545 this should implement a solution to it!

**Review by KATT — COMMENTED** [review]
@juliusmarminge can you review this please? you're the master on this :D

**Review by juliusmarminge — APPROVED** [review]
Looks good but I think we can clean it up a bit on `next` where queryKeys aren't this messy...

**Comment by SrBrahma** [comment]
@juliusmarminge before / if you are going to merge it, I suggest you read this discussion I participated in react-query after I made this PR:

https://github.com/TanStack/query/discussions/4762#discussioncomment-4605555

Tl;dr: I don't know if this PR is good as if a query is enabled=false, the query params may have invalid values types.

Maybe we could have the type to be something like `{ enabled: false, queryKey: [x, any] } | { enabled?: true, queryKey: [x, theRightType] }`?

While this PR works, it may be dangerous if enabled is false as it may potentially lead to invalid data access.

**Comment by juliusmarminge** [comment]
> Maybe we could have the type to be something like ~ { enabled: false, queryKey: [x, unknown] } | { enabled?: true, queryKey: [x, theRightType] }, so the user will need to check for enabled === undefined || enabled === true to then use the data with the right type?

This wont matter since the input type will be a DeepPartial so you have to check for every prop you try to access no matter what

**Review by juliusmarminge — DISMISSED** [review]

**Comment by nicolassanmar** [comment]
@juliusmarminge @SrBrahma 
> While this PR works, it may be dangerous if enabled is false as it may potentially lead to invalid data access.

I think that typing the predicate still makes sense, as it will be right in regular use cases. People can only find data in the cache with unexpected typings if:
1. If they tell TS to ignore a typing error (for example, a non null assertion, even when `enabled: false`).
1. If they tampered with the react-query cache in some way

Somewhere in the documentation should state that the typing is only true if they did not do any of the above.

**Review by juliusmarminge — APPROVED** [review]
I think this is fine. Since the type is a deep partial, users will still need to check for nested property accesses so that shouldn't be an issue, no matter if `enabled: false` is used or not(?).

Maybe @TkDodo has some insights too?

**Review by juliusmarminge — APPROVED** [review]

**Comment by TkDodo** [comment]
> I think this is fine. Since the type is a deep partial, users will still need to check for nested property accesses so that shouldn't be an issue, no matter if enabled: false is used or not(?).

> Maybe @TkDodo has some insights too?

I re-checked the linked react-query discussion, and I think I understand the issue now better. Note that queries where all observers are disabled are explicitly exempt from being refetched, but the predicate filtering is still applied first:

- https://github.com/TanStack/query/blob/765f6196e900c1dd826b040773fcf97839ca60d0/packages/query-core/src/queryClient.ts#L253-L256

If the type will be deep-partial in the future, I think this will side-step the issue 👍

**Review by KATT — APPROVED** [review]
